ec2/elasticip.py

```diff
@@ -14,109 +14,123 @@
 #    * limitations under the License.
 
 # Boto Imports
-from boto.exception import EC2ResponseError
-from boto.exception import BotoServerError
+import boto.exception
 
 # Cloudify imports
 from cloudify import ctx
 from cloudify.exceptions import NonRecoverableError
 from cloudify.decorators import operation
 from ec2 import connection
+from ec2 import utils
 
 
 @operation
-def allocate(**kwargs):
+def allocate(**_):
     """ Allocates an Elastic IP in the connected region in the AWS account.
     """
     ec2_client = connection.EC2ConnectionClient().client()
     ctx.logger.info('Allocating Elastic IP.')
 
+    if ctx.node.properties['use_external_resource']:
+        address = utils.get_address_by_id(
+            ctx.node.properties.get('resource_id'), ctx=ctx)
+        ctx.instance.runtime_properties['aws_resource_id'] = address
+        ctx.logger.info('Using existing resource: {0}'.format(address))
+        return
+
     try:
-        address_object = ec2_client.allocate_address()
-    except (EC2ResponseError, BotoServerError) as e:
-        raise NonRecoverableError('Error. Failed to '
-                                  'provision Elastic IP. Error: {0}.'
-                                  .format(e))
+        address_object = ec2_client.allocate_address(domain=None)
+    except (boto.exception.EC2ResponseError,
+            boto.exception.BotoServerError) as e:
+        raise NonRecoverableError('Failed to provision Elastic IP. Error: {0}.'
```

Comment:
It's not idempotent.
I would propose that we do a test to make sure that it was provisioned (which would be slow) rather than allow it to be retried.